From 89e009b5fe0e65f34fda0fe83dbbe691f7b76b59 Mon Sep 17 00:00:00 2001
From: Venkateshwar Rao Gannavarapu <venkateshwar.rao.gannavarapu@xilinx.com>
Date: Wed, 25 Sep 2019 10:33:39 -0700
Subject: [PATCH 0672/1851] drm: xlnx: mixer: Add a delay for timing generation
 to be stable

commit 57808444dab17da0b01e38679df13c7b3bd19a49 from
https://github.com/Xilinx/linux-xlnx.git

This patch adds vblank processing delay about 3 frame periods to avoid
timeout in processing the first frame. Without this patch, vblank
timeout warn message is observed for the first frame.

Signed-off-by: Venkateshwar Rao Gannavarapu <venkateshwar.rao.gannavarapu@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_mixer.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_mixer.c b/drivers/gpu/drm/xlnx/xlnx_mixer.c
index c459c18976c6..82b3f151c93c 100644
--- a/drivers/gpu/drm/xlnx/xlnx_mixer.c
+++ b/drivers/gpu/drm/xlnx/xlnx_mixer.c
@@ -2545,7 +2545,15 @@ static void
 xlnx_mix_crtc_atomic_enable(struct drm_crtc *crtc,
 			    struct drm_crtc_state *old_crtc_state)
 {
+	struct drm_display_mode *adjusted_mode = &crtc->state->adjusted_mode;
+	int vrefresh;
+
 	xlnx_mix_crtc_dpms(crtc, DRM_MODE_DPMS_ON);
+
+	/* Delay of 3 vblank interval for timing gen to be stable */
+	vrefresh = ((adjusted_mode->clock * 1000) /
+		    (adjusted_mode->vtotal * adjusted_mode->htotal));
+	msleep(3 * 1000 / vrefresh);
 }
 
 /**
-- 
2.31.1

